crates/common/consensus/src/deneb/beacon_state.rs

```diff
@@ -78,7 +88,58 @@ pub struct BeaconState {
 }
 
 impl BeaconState {
+    /// Return the current epoch.
     pub fn get_current_epoch(&self) -> u64 {
         compute_epoch_at_slot(self.slot)
     }
+
+    /// Return the previous epoch (unless the current epoch is ``GENESIS_EPOCH``).
+    pub fn get_previous_epoch(&self) -> u64 {
+        let current_epoch = self.get_current_epoch();
+        if current_epoch == GENESIS_EPOCH {
+            GENESIS_EPOCH
+        } else {
+            current_epoch - 1
+        }
+    }
+
+    /// Return the block root at the start of a recent ``epoch``.
+    pub fn get_block_root(&self, epoch: u64) -> B256 {
+        self.get_block_root_at_slot(compute_start_slot_at_epoch(epoch))
+    }
+
+    /// Return the block root at a recent ``slot``.
+    pub fn get_block_root_at_slot(&self, slot: u64) -> B256 {
+        assert!(slot < self.slot && self.slot <= slot + SLOTS_PER_HISTORICAL_ROOT);
+        self.block_roots[(slot % SLOTS_PER_HISTORICAL_ROOT) as usize]
+    }
```

Comment:
```suggestion
    /// Return the block root at the start of a recent ``epoch``.
    pub fn get_block_root(&self, epoch: u64) -> anyhow::Result<B256> {
        self.get_block_root_at_slot(compute_start_slot_at_epoch(epoch))
    }

    /// Return the block root at a recent ``slot``.
    pub fn get_block_root_at_slot(&self, slot: u64) -> anyhow::Result<B256> {
        bail!(slot < self.slot && self.slot <= slot + SLOTS_PER_HISTORICAL_ROOT, "slot given was outside of block_roots range");
        Ok(self.block_roots[(slot % SLOTS_PER_HISTORICAL_ROOT) as usize])
    }
```

use bail instead of an assert, so the program doesn't crash and we can peacefully handle the error